### ISSUE TITLE
Triton yolo predict imgsz issue

### ISSUE THREAD

**Issue by ankhafizov** [ultralytics/ultralytics#18320]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

```
from ultralytics import YOLO

model = YOLO(f"http://{triton_socket_http}/{model_name}", task="classify")
results = model.predict(frame, imgsz=(224,416), verbose=False)
```

gives following error:

unexpected shape for input 'images' for model 'anomaly_classifier_leak'. Expected [-1,3,224,416], got [1,3,224,224]. NOTE: Setting a non-zero max_batch_size in the model config requires a batch dimension to be prepended to each input shape. If you want to specify the full shape including the batch dim in your input dims config, try setting max_batch_size to zero. See the model configuration docs for more info on max_batch_size.

It is because it takes size as 224, nevertheless I gave (224,416). On triton side I have converted to onnx yolov8 model

### Environment

Ultralytics 8.3.49 🚀 Python-3.11.8 torch-2.2.2+cu121 CUDA:0 (NVIDIA GeForce RTX 3080 Ti Laptop GPU, 16384MiB)
Setup complete ✅ (20 CPUs, 15.5 GB RAM, 174.2/1006.9 GB disk)

OS                  Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
Environment         Docker
Python              3.11.8
Install             pip
RAM                 15.47 GB
Disk                174.2/1006.9 GB
CPU                 12th Gen Intel Core(TM) i9-12900H
CPU count           20
GPU                 NVIDIA GeForce RTX 3080 Ti Laptop GPU, 16384MiB
GPU count           1
CUDA                12.1

numpy               ✅ 1.26.4>=1.23.0
numpy               ✅ 1.26.4<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.9.2>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.3.0>=7.1.2
pyyaml              ✅ 6.0.1>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.13.0>=1.4.1
torch               ✅ 2.2.2+cu121>=1.8.0
torch               ✅ 2.2.2+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.17.2+cu121>=0.9.0
tqdm                ✅ 4.66.2>=4.64.0
psutil              ✅ 5.9.8
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
{'OS': 'Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35',
 'Environment': 'Docker',
 'Python': '3.11.8',
 'Install': 'pip',
 'RAM': '15.47 GB',
 'Disk': '174.2/1006.9 GB',
 'CPU': '12th Gen Intel Core(TM) i9-12900H',
 'CPU count': 20,
 'GPU': 'NVIDIA GeForce RTX 3080 Ti Laptop GPU, 16384MiB',
 'GPU count': 1,
 'CUDA': '12.1',
 'Package Info': {'numpy': '✅ 1.26.4<2.0.0; sys_platform == "darwin"',
  'matplotlib': '✅ 3.9.2>=3.3.0',
  'opencv-python': '✅ 4.10.0.84>=4.6.0',
  'pillow': '✅ 10.3.0>=7.1.2',
  'pyyaml': '✅ 6.0.1>=5.3.1',
  'requests': '✅ 2.31.0>=2.23.0',
  'scipy': '✅ 1.13.0>=1.4.1',
  'torch': '✅ 2.2.2+cu121!=2.4.0,>=1.8.0; sys_platform == "win32"',
  'torchvision': '✅ 0.17.2+cu121>=0.9.0',
  'tqdm': '✅ 4.66.2>=4.64.0',
  'psutil': '✅ 5.9.8',
  'py-cpuinfo': '✅ 9.0.0',
  'pandas': '✅ 2.2.2>=1.1.4',
  'seaborn': '✅ 0.13.2>=0.11.0',
  'ultralytics-thop': '✅ 2.0.13>=2.0.0'}}

### Minimal Reproducible Example

```
from ultralytics import YOLO

model = YOLO(f"http://{triton_socket_http}/{model_name}", task="classify")
results = model.predict(frame, imgsz=(224,416), verbose=False)
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ankhafizov, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug it. This will allow us to better understand the issue and pinpoint the cause. From your code snippet, it seems you're running into an input shape mismatch while using Triton. Please ensure your model configuration aligns with your usage of `imgsz`. If the issue persists, providing any additional configuration details or logs can also help us troubleshoot 🛠️.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will follow up with you shortly to assist further! 😊

**Comment by Y-T-G** [comment]
What's the code you used to export it? Did you export with `imgsz=224`?

**Comment by ankhafizov** [comment]
No, 224x416

> What's the code you used to export it? Did you export with `imgsz=224`?

**Comment by Y-T-G** [comment]
Why? Classification models are trained with square imgsz.

**Comment by Y-T-G** [comment]
Does the ONNX model work when you load it in Ultralytics?

It's not an issue with Triton. Classification models only support square image inference.

**Comment by ankhafizov** [comment]
> Why? Classification models are trained with square imgsz.

Because the standart inference only see the center area of the image (yolo does CenterCrop)

**Comment by ankhafizov** [comment]
> Does the ONNX model work when you load it in Ultralytics?
> 
> It's not an issue with Triton. Classification models only support square image inference in Ultralytics

Im yolo documentation they say, that it is pissible to provide list for imgsz. Standart yolo inference works with the exported model.

Even I can inference the yolo triton service using self-written methods:

```
def infer_triton_yolo_classifier(
    triton_client,
    triton_model_name,
    image,
    imgsz,
    nclasses,
    input_name="images",
    output_name="output0",
) -> dict[str, float]:
    assert isinstance(imgsz, Iterable) and len(imgsz)==2, "imgsz - должен быть список из 2 чисел (h, w)"
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)

    image = cv2.resize(image, (imgsz[1], imgsz[0]), interpolation=cv2.INTER_AREA)
    image = image.astype(np.float32) / 255
    image = np.transpose(image, (2, 0, 1))  # Convert to CHW format

    image = np.expand_dims(image, axis=0)

    inputs = [httpclient.InferInput(input_name, image.shape, "FP32")]
    inputs[0].set_data_from_numpy(image)
    outputs = [
        httpclient.InferRequestedOutput(output_name, binary_data=False, class_count=nclasses)
    ]
    results = triton_client.infer(triton_model_name, inputs, outputs=outputs)
    results: list[str] = results.as_numpy(output_name)[0]

    probas = {}
    for res_item in results:
        proba, _, cls = res_item.strip().split(":")
        probas[cls] = float(proba)

    return probas
```

**Comment by glenn-jocher** [comment]
Thank you for sharing your method. While YOLO models support non-square image sizes in general, classification tasks in Ultralytics enforce square resizing during preprocessing to maintain consistency with model training. This is likely the source of the discrepancy. If your custom Triton service works as expected for your use case, you can continue leveraging it. For troubleshooting or additional flexibility, you may consider modifying the preprocessing pipeline in the exported ONNX model or adapting your workflow accordingly.

**Comment by Y-T-G** [comment]
> Im yolo documentation they say, that it is pissible to provide list for imgsz.

It works for other models, not classification. Classification only supports square image inference.

> Standart yolo inference works with the exported model.

What's the image size shown in logs during inference with the ONNX model?

> Because the standart inference only see the center area of the image (yolo does CenterCrop)

It will still use CenterCrop even if you export with non-square imgsz because Ultralytics only supports square image size for classification. It will use the first value of the imgsz you pass. So if you pass `imgsz=(224, 416)` in `model.predict()`, it will resize to `(224, 224)` and if you pass `imgsz=(416, 224)`, it will resize to `(416, 416)`. You can check the prediction logs to verify that.

```python
In [4]: model = YOLO("yolo11n-cls.onnx")

In [5]: results = model.predict(ASSETS / "bus.jpg"
   ...: , imgsz=(224, 416))
Loading yolo11n-cls.onnx for ONNX Runtime inference...
Preferring ONNX Runtime AzureExecutionProvider

image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 224x224 minibus 0.57, police_van 0.34, trolleybus 0.04, recreational_vehicle 0.01, streetcar 0.01, 29.3ms
Speed: 14.9ms preprocess, 29.3ms inference, 0.2ms postprocess per image at shape (1, 3, 224, 224)
```

It resizes to `(224, 224)` as shown in the logs, even if you pass `imgsz=(224, 416)`.

**Comment by Y-T-G** [comment]
If you want non-square imgsz, then you will have to perform preprocessing manually and then pass `torch.tensor` (with shape `1, 3, 224, 416`) as input to `model.predict()` instead of numpy array.

**Comment by ankhafizov** [comment]
> Thank you for sharing your method. While YOLO models support non-square image sizes in general, classification tasks in Ultralytics enforce square resizing during preprocessing to maintain consistency with model training. This is likely the source of the discrepancy. If your custom Triton service works as expected for your use case, you can continue leveraging it. For troubleshooting or additional flexibility, you may consider modifying the preprocessing pipeline in the exported ONNX model or adapting your workflow accordingly.

so, the classification models performs letterbox transform or centercrop to make the image square?

**Comment by ankhafizov** [comment]
> > Im yolo documentation they say, that it is pissible to provide list for imgsz.
> 
> It works for other models, not classification. Classification only supports square image inference.
> 
> > Standart yolo inference works with the exported model.
> 
> What's the image size shown in logs during inference with the ONNX model?
> 
> > Because the standart inference only see the center area of the image (yolo does CenterCrop)
> 
> It will still use CenterCrop even if you export with non-square imgsz because Ultralytics only supports square image size for classification. It will use the first value of the imgsz you pass. So if you pass `imgsz=(224, 416)` in `model.predict()`, it will resize to `(224, 224)` and if you pass `imgsz=(416, 224)`, it will resize to `(416, 416)`. You can check the prediction logs to verify that.
> 
> ```python
> In [4]: model = YOLO("yolo11n-cls.onnx")
> 
> In [5]: results = model.predict(ASSETS / "bus.jpg"
>    ...: , imgsz=(224, 416))
> Loading yolo11n-cls.onnx for ONNX Runtime inference...
> Preferring ONNX Runtime AzureExecutionProvider
> 
> image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 224x224 minibus 0.57, police_van 0.34, trolleybus 0.04, recreational_vehicle 0.01, streetcar 0.01, 29.3ms
> Speed: 14.9ms preprocess, 29.3ms inference, 0.2ms postprocess per image at shape (1, 3, 224, 224)
> ```
> 
> It resizes to `(224, 224)` as shown in the logs, even if you pass `imgsz=(224, 416)`.

```
from ultralytics import YOLO

img_size = (224, 416)
empty_video_pth = "videos/Sakhara_83_18_12_2024 07.17.40.mp4"

yolo_model = YOLO("weights/empty.onnx", task="classify")
yolo_model.predict(empty_video_pth)
```

![image](https://github.com/user-attachments/assets/acf7d874-62d2-499f-96dc-23d595589ab7)

**Comment by Y-T-G** [comment]
It says `640x640` in the logs

**Comment by ankhafizov** [comment]
> > Im yolo documentation they say, that it is pissible to provide list for imgsz.
> 
> It works for other models, not classification. Classification only supports square image inference.
> 
> > Standart yolo inference works with the exported model.
> 
> What's the image size shown in logs during inference with the ONNX model?
> 
> > Because the standart inference only see the center area of the image (yolo does CenterCrop)
> 
> It will still use CenterCrop even if you export with non-square imgsz because Ultralytics only supports square image size for classification. It will use the first value of the imgsz you pass. So if you pass `imgsz=(224, 416)` in `model.predict()`, it will resize to `(224, 224)` and if you pass `imgsz=(416, 224)`, it will resize to `(416, 416)`. You can check the prediction logs to verify that.
> 
> ```python
> In [4]: model = YOLO("yolo11n-cls.onnx")
> 
> In [5]: results = model.predict(ASSETS / "bus.jpg"
>    ...: , imgsz=(224, 416))
> Loading yolo11n-cls.onnx for ONNX Runtime inference...
> Preferring ONNX Runtime AzureExecutionProvider
> 
> image 1/1 /ultralytics/ultralytics/assets/bus.jpg: 224x224 minibus 0.57, police_van 0.34, trolleybus 0.04, recreational_vehicle 0.01, streetcar 0.01, 29.3ms
> Speed: 14.9ms preprocess, 29.3ms inference, 0.2ms postprocess per image at shape (1, 3, 224, 224)
> ```
> 
> It resizes to `(224, 224)` as shown in the logs, even if you pass `imgsz=(224, 416)`.

```python
from ultralytics import YOLO
import cv2
from time import sleep

img_size = (224, 416)
empty_video_pth = "videos/Sakhara_83_18_12_2024 07.17.40.mp4"

yolo_model = YOLO("weights/empty.onnx", task="classify")
yolo_model.predict(empty_video_pth, imgsz=(224, 416))
```

no size warnings((

![image](https://github.com/user-attachments/assets/f9c3fd1c-fb37-48f3-90d6-383304ef0161)

**Comment by Y-T-G** [comment]
> > Does the ONNX model work when you load it in Ultralytics?
> > It's not an issue with Triton. Classification models only support square image inference in Ultralytics
> 
> Im yolo documentation they say, that it is pissible to provide list for imgsz. Standart yolo inference works with the exported model.
> 
> Even I can inference the yolo triton service using self-written methods:
> 
> ```
> def infer_triton_yolo_classifier(
>     triton_client,
>     triton_model_name,
>     image,
>     imgsz,
>     nclasses,
>     input_name="images",
>     output_name="output0",
> ) -> dict[str, float]:
>     assert isinstance(imgsz, Iterable) and len(imgsz)==2, "imgsz - должен быть список из 2 чисел (h, w)"
>     image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
> 
>     image = cv2.resize(image, (imgsz[1], imgsz[0]), interpolation=cv2.INTER_AREA)
>     image = image.astype(np.float32) / 255
>     image = np.transpose(image, (2, 0, 1))  # Convert to CHW format
> 
>     image = np.expand_dims(image, axis=0)
> 
>     inputs = [httpclient.InferInput(input_name, image.shape, "FP32")]
>     inputs[0].set_data_from_numpy(image)
>     outputs = [
>         httpclient.InferRequestedOutput(output_name, binary_data=False, class_count=nclasses)
>     ]
>     results = triton_client.infer(triton_model_name, inputs, outputs=outputs)
>     results: list[str] = results.as_numpy(output_name)[0]
> 
>     probas = {}
>     for res_item in results:
>         proba, _, cls = res_item.strip().split(":")
>         probas[cls] = float(proba)
> 
>     return probas
> ```

You can use this same code for preprocessing and then pass the torch tensor.

```
image = torch.from_numpy(image)
results = model.predict(image)
```

This will skip Ultralytics preprocessing and use your preprocessed tensor directly.

**Comment by ankhafizov** [comment]
> > > Does the ONNX model work when you load it in Ultralytics?
> > > It's not an issue with Triton. Classification models only support square image inference in Ultralytics
> > 
> > 
> > Im yolo documentation they say, that it is pissible to provide list for imgsz. Standart yolo inference works with the exported model.
> > Even I can inference the yolo triton service using self-written methods:
> > ```
> > def infer_triton_yolo_classifier(
> >     triton_client,
> >     triton_model_name,
> >     image,
> >     imgsz,
> >     nclasses,
> >     input_name="images",
> >     output_name="output0",
> > ) -> dict[str, float]:
> >     assert isinstance(imgsz, Iterable) and len(imgsz)==2, "imgsz - должен быть список из 2 чисел (h, w)"
> >     image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
> > 
> >     image = cv2.resize(image, (imgsz[1], imgsz[0]), interpolation=cv2.INTER_AREA)
> >     image = image.astype(np.float32) / 255
> >     image = np.transpose(image, (2, 0, 1))  # Convert to CHW format
> > 
> >     image = np.expand_dims(image, axis=0)
> > 
> >     inputs = [httpclient.InferInput(input_name, image.shape, "FP32")]
> >     inputs[0].set_data_from_numpy(image)
> >     outputs = [
> >         httpclient.InferRequestedOutput(output_name, binary_data=False, class_count=nclasses)
> >     ]
> >     results = triton_client.infer(triton_model_name, inputs, outputs=outputs)
> >     results: list[str] = results.as_numpy(output_name)[0]
> > 
> >     probas = {}
> >     for res_item in results:
> >         proba, _, cls = res_item.strip().split(":")
> >         probas[cls] = float(proba)
> > 
> >     return probas
> > ```
> 
> You can use this same code for preprocessing and then pass the torch tensor.
> 
> ```
> image = torch.from_numpy(image)
> results = model.predict(image)
> ```
> 
> This will skip Ultralytics preprocessing and use your preprocessed tensor directly.

seems like that, thank you!

**Comment by glenn-jocher** [comment]
You're welcome! If you encounter any further issues or have additional questions, feel free to ask. The YOLO and Ultralytics communities are always here to help!

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in the linked PR (#16630) by Y-T-G, with valuable contributions from Laughing-q and glenn-jocher.

Key improvements from this update include:

- Improved image classification transforms to ensure correct preprocessing for non-square image sizes.
- Enhanced dependency handling in the similarity search feature, reducing import-related errors.
- More consistent image size management by leveraging model export metadata when available.
- Updated the Ultralytics package version to 8.3.138.

These enhancements are designed to provide more reliable classification results and a smoother experience with Ultralytics models.

To test if these changes resolve your issue, please upgrade to the latest version using:
```
pip install -U ultralytics>=8.3.138
```
Alternatively, you can install the latest code directly from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
You can also wait for the next official PyPI release if you prefer.

Please let us know if the update addresses your concerns or if you encounter any further issues. Thanks for reporting this, and any additional feedback is always welcome! 🙏
